application/models/services/SurveyAccessModeService.php

```diff
@@ -0,0 +1,183 @@
+        $oldTable = "tokens_" . $survey->sid;
+        $newTable = "old_tokens_" . $survey->sid . "_" . $date;
+        $userID = $this->app->user->getId();
+
+        if ($survey->active !== 'Y') {
+            return;
+        }
+
+        if ($survey->hasTokensTable) {
+            if (!in_array($action, self::$supportedActions)) {
+                $action = self::$ACTION_KEEP;
+            }
+            $tokenSample = Token::model($survey->sid)->find('1=1');
+            if ($tokenSample === null) {
+                $action = self::$ACTION_DROP;
+            }
+            if ($action === self::$ACTION_ARCHIVE) {
+                $surveyInfo = getSurveyInfo($survey->sid);
+                $this->app->db->createCommand()->renameTable("{{" . $oldTable . "}}", "{{" . $newTable . "}}");
+                $archivedTokenSettings = new ArchivedTableSettings();
+                $archivedTokenSettings->survey_id = $survey->sid;
+                $archivedTokenSettings->user_id = $userID;
+                $archivedTokenSettings->tbl_name = $newTable;
+                $archivedTokenSettings->tbl_type = 'token';
+                $archivedTokenSettings->created = $DBDate;
+                $archivedTokenSettings->properties = $surveyInfo['tokenencryptionoptions'];
+                $archivedTokenSettings->attributes = json_encode($surveyInfo['attributedescriptions']);
+                $archivedTokenSettings->save();
+            } elseif ($action === self::$ACTION_DROP) {
+                $this->app->db->createCommand()->dropTable("{{" . $oldTable . "}}");
+            } //If action is Keep, do nothing
+        }
+    }
+
+    /**
+     * Changes the access mode of the survey
+     * @param int $surveyID the id of the survey whose access mode is to be changed
+     * @param string $accessMode the access mode we desire to have
+     * @param string $action whether we intend to archive the tokens table or not
+     * @throws \LimeSurvey\Models\Services\Exception\PersistErrorException
+     * @throws \LimeSurvey\Models\Services\Exception\PermissionDeniedException
+     * @return bool whether the change was done
+     */
+    public function changeAccessMode(int $surveyID, string $accessMode, string $action = 'K')
+    {
+        $survey = Survey::model()->findByPk($surveyID);
+        $oldAccessMode = $survey->access_mode;
+        if ($oldAccessMode === $accessMode) {
+            return false; //Nothing to change
+        }
+        if (!in_array($accessMode, self::$supportedAccessModes)) {
+            throw new PersistErrorException(
+                'The access mode given is not supported'
+            );
+        }
+        if (!$this->hasPermission($surveyID, $accessMode)) {
+            throw new PermissionDeniedException(
+                'Access denied'
+            );
+        }
+        $survey->access_mode = $accessMode;
+        if ($oldAccessMode === self::$ACCESS_TYPE_OPEN) {
+            $this->newTokenTable($survey);
```

Comment:
We shouldrename this to 'newParticipantTable' etc. to communicate the current terminology.